Optmized data sent during RPC and RSet calls.

AndreaCatania opened the pull request:
- Now is sent the method ID rather the full function name.
- The passed IDs (Node and Method) are compressed so to use less possible space.
- The variant (INT and BOOL) is now encoded and compressed so to use much less data.
- Optimized RPCMode retrieval for GDScript functions.

This work has been kindly sponsored by IMVU.

fire commented on modules/visual_script/visual_script.cpp:
```diff
@@ -1206,6 +1260,30 @@ void VisualScript::_set_data(const Dictionary &p_data) {
+		}
+	}
+
+	// Visual script doesn't have rset :(
```
:( How do I get this in?

AndreaCatania replied:
oh, You have to add the class variables first because these are totally missing

AndreaCatania commented:
Fixed the CI builds.

I did a tests to measure the amount of data transferred each sec:

```gdscript
$"../World/Node2".rpc("test_string_rpc", "String test stest stea")
$"../World/Node2".rpc("test_bool_rpc", true)
$"../World/Node2".rpc("test_int_rpc", 120)
$"../World/Node2".rpc("test_float_rpc", 120.3)
$"../World/Node2".rpc("test_int_array_rpc", [12, 13, 14])
```
I'm sending `bool / int / string / floats / array`
- Current master: __11.66Kb/s__
- This PR: __5.8 Kb/s__

Rough 50% less data transferred.

Zireael07 commented:
Would this also improve the amount of data transferred to in-editor debugger?

AndreaCatania commented:
> Would this also improve the amount of data transferred to in-editor debugger?

Unfortunately nope, this works only for the `MultiplayerAPI`.

akien-mga commented:
CC @Faless

Faless commented on core/io/multiplayer_api.cpp:
```diff
@@ -468,6 +518,154 @@ bool MultiplayerAPI::_send_confirm_path(NodePath p_path, PathSentCache *psc, int
+
+	return OK;
+}
+Error MultiplayerAPI::_decode_and_decompress_variant(Variant &r_variant, const uint8_t *p_buffer, int p_len, int *r_len) {
```
This should probably find its way into marshalls code in `core/io/marshalls.h`

Faless commented on core/io/multiplayer_api.cpp:
```diff
@@ -468,6 +518,154 @@ bool MultiplayerAPI::_send_confirm_path(NodePath p_path, PathSentCache *psc, int
+#define ENCODE_16 1 << 5
+#define ENCODE_32 2 << 5
+#define ENCODE_64 3 << 5
+Error MultiplayerAPI::_encode_and_compress_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len) {
```
This should probably find its way into marshalls code in `core/io/marshalls.h`

Faless commented on scene/main/node.cpp:
```diff
@@ -498,22 +500,38 @@ bool Node::is_network_master() const {
+		nd.name = p_method;
+		nd.mode = p_mode;
+		data.rpc_methods.push_back(nd);
+		return ((uint16_t)data.rpc_properties.size() - 1) | (1 << 15);
```
Does this mean different `rpc_config` call order will cause the ID to be different? Shouldn't we just sort the IDs again?

Faless commented on scene/main/node.cpp:
```diff
@@ -498,22 +500,38 @@ bool Node::is_network_master() const {
 }
 
-void Node::rset_config(const StringName &p_property, MultiplayerAPI::RPCMode p_mode) {
+uint16_t Node::rset_config(const StringName &p_property, MultiplayerAPI::RPCMode p_mode) {
```
See above:
Does this mean different `rset_config` call order will cause the ID to be different? Shouldn't we just sort the IDs again?

Faless reviewed (changes requested):
The work overall looks really good, and it's going to be a great improvement.

Some things worries me though:

1. This will break workflow for people who uses different scripts for client and server. Unless the script have the same exposed RPC/RSET (order shouldn't matter since sorting is done based on case). This is okay in general, but it needs to fail properly. Maybe we can spend a byte to do a checksum (xors of function names/...) so it's detectable and we can report it.

2. As pointed out in comments, if I understood correctly the code, exposing a new method via rpc_config will just append it. This means that different rpc_config order will result in differnt ids being assigned, and the wrong function to be called. Shouldn't we re-sort them when that happens? That will fix that problem, but ongoing calls with that ID will fail instead. For now, `rpc_config` seems to crash the engine when the new RPC is sent (see attached code).

3. The encoding optimization should probably be moved to marshalls, but I guess that could be done in a later PR.

Of all the points, `2` worries me the most. I feel we have to find a solution to that. Do you have anything in mind (maybe some form of versioning?). Using an agreed upon ID like we do for paths is less optimized and it would be a pity to waste the work done so far on generating the ID.

The folling code crash the engine (rpc_config):
```gdscript
extends Node

func _ready():
	var args = Array(OS.get_cmdline_args())
	while args.size():
		var a = args.pop_front()
		if a == 'c':
			var peer = NetworkedMultiplayerENet.new()
			peer.create_client("localhost", 8092)
			multiplayer.connect("network_peer_connected", self, "_connected")
			multiplayer.connect("network_peer_packet", self, "_pkt")
			multiplayer.set_network_peer(peer)
		elif a == 's':
			var peer = NetworkedMultiplayerENet.new()
			peer.create_server(8092)
			multiplayer.connect("network_peer_connected", self, "_connected")
			multiplayer.connect("network_peer_packet", self, "_pkt")
			multiplayer.set_network_peer(peer)
	rpc_config("_added", MultiplayerAPI.RPC_MODE_REMOTESYNC)
	rpc_config("_other", MultiplayerAPI.RPC_MODE_REMOTESYNC)

func _pkt(p_id : int, pkt : PoolByteArray):
	if pkt[0] != 0:
		return # Just for testing
	if not multiplayer.is_network_server():
		prints(p_id, pkt.subarray(1, pkt.size() - 1).get_string_from_utf8(), "at", OS.get_ticks_msec())
		multiplayer.send_bytes(pkt, 1)
	else:
		pkt = pkt.subarray(1, pkt.size() - 1)
		print("Recv ping at: %s = %s" % [OS.get_ticks_msec(), OS.get_ticks_msec() - int(pkt.get_string_from_utf8())])
		return
		multiplayer.send_bytes(str(OS.get_ticks_msec()).to_utf8(), p_id)

func _connected(p_id):
	print("Connected to: %d" % p_id)
	if multiplayer.is_network_server():
		while true:
			yield(get_tree().create_timer(1),"timeout")
			if not p_id in multiplayer.get_network_connected_peers():
				break
			print("Sending ping at ", OS.get_ticks_msec())
			var pkt = PoolByteArray([0])
			pkt.append_array(str(OS.get_ticks_msec()).to_utf8())
			multiplayer.send_bytes(pkt, p_id)
			rpc_id(p_id, "_ping", OS.get_ticks_msec())
			rpc("_added", "Added Called") #### !!!!!!!!!!! CRASH HERE.
			rpc("_other", "Other Called")

	if multiplayer.is_network_server():
		var ticks = OS.get_ticks_msec()
		print("Ping: %d" % ticks)
		rpc_id(p_id, "_ping", ticks)
		for i in range(0, 10):
			yield(get_tree().create_timer(1),"timeout")
			ticks = OS.get_ticks_msec()
			print("Ping: %d" % ticks)
			rpc_id(p_id, "_ping", ticks)

remote func _ping(time):
	print("Ping: %d" % time)
	rpc_id(1, "_pong", time)

remote func _pong(time):
	print("Pong: %d - %d = %d" % [OS.get_ticks_msec(), time, OS.get_ticks_msec() - time])

func _added(msg):
	print("Added called with: ", msg)
	pass
func _other(msg):
	print("Other called with: ", msg)
	pass
```

Reminder for self: Double check int/size comparisons.

AndreaCatania commented:
I've solved the crash, it was a too silly bug :/ ... 

However,

> 1. This will break workflow for people who uses different scripts for client and server. Unless the script have the same exposed RPC/RSET (order shouldn't matter since sorting is done based on case). This is okay in general, but it needs to fail properly. Maybe we can spend a byte to do a checksum (xors of function names/...) so it's detectable and we can report it.

If we really want such a check we can add a mechanism that ensure that the rpcs method names are the same across the peers, during the node name caching phase so we can save bits in sub sequent calls; Also we can disable this check in release so to avoid useless check.


> 2. As pointed out in comments, if I understood correctly the code, exposing a new method via rpc_config will just append it. This means that different rpc_config order will result in differnt ids being assigned, and the wrong function to be called. Shouldn't we re-sort them when that happens? That will fix that problem, but ongoing calls with that ID will fail instead. For now, rpc_config seems to crash the engine when the new RPC is sent (see attached code).

I prefer to not sort that functions by name because the `rpc_config` is supposed to be called always with the same order anywhere. Also. with the above check we can handle the fail condition correctly.

> 3. The encoding optimization should probably be moved to marshalls, but I guess that could be done in a later PR.

Initially I added a boolean to the marshal function to activate padding removal but Juan told me to keep it otside the marshalling; so I moved the code where it's now. I've to admit that this solution is much more elegant compared to the previous one.

AndreaCatania committed:
Optmized data sent during RPC and RSet calls.
- Now is sent the method ID rather the full function name.
- The passed IDs (Node and Method) are compressed so to use less possible space.
- The variant (INT and BOOL) is now encoded and compressed so to use much less data.
- Optimized RPCMode retrieval for GDScript functions.
- Added checksum to assert the methods are the same across peers.

This work has been kindly sponsored by IMVU.

AndreaCatania commented:
According to @Faless I've added a check to ensure both the server and the clients have the same methods IDs.

Rebased and make it ready for merge.

akien-mga commented:
Are documentation updates in the classref and online tutorials needed, or is this all internal optimization transparent to the user?

AndreaCatania commented:
This is transparent optimization, nothing changed from the user prospective.

Faless reviewed (approved):
Looking good to me. Great job :trophy:

akien-mga commented:
Thanks!

shinyclaw commented:
Can we merge this commit to 3.2 branch?

Faless commented:
> Can we merge this commit to 3.2 branch?

@shinyclaw No, since it will break games that rely on having different scripts for server and clients, in cases where method names and/or number differs among the scripts.

shinyclaw commented:
Thanks Faless - look like I have exactly the same situation on my project (different scripts for server and clients), but even in 3.2 I must have the same method defined on server and clients (as empty ones with name and pass keyword). If not, when clients send RPC over the network, the server throws errors like that:

ERROR: RPC 'change_weapon' is not allowed on node /root/World/Players/1652423913 from: 1652423913. Mode is 0, master is 1652423913.
   At: core/io/multiplayer_api.cpp:291
